cmake/compiler/clang/target.cmake: Improve support for different runtime libraries

duda-patryk opened the pull request:
Use rtlib provided by compiler instead of hardcoding libgcc

Clang provides runtime library `libclang_rt.builtins.<arch>.a` but `libgcc.a` is also supported.
The compiler can provide full path to the selected library using the `--print-libgcc-file-name` option, for example:
```
/usr/lib64/clang/17/lib/baremetal/libclang_rt.builtins-armv7m.a
```

If `--rtlib=libgcc` option is provided, clang will also provide appropriate path.

Replace hardcoded `libgcc` with library name extracted from path, so we are compatible with both libraries.

duda-patryk committed:
clang: Use rtlib provided by compiler instead of hardcoding libgcc

Clang provides runtime library `libclang_rt.builtins.<arch>.a` but
`libgcc.a` is also supported.

The compiler can provide full path to the selected library using the
`--print-libgcc-file-name` option, for example:
```
/usr/lib64/clang/17/lib/baremetal/libclang_rt.builtins-armv7m.a
```

If `--rtlib=libgcc` option is provided, clang will also provide
appropriate path.

This patch replaces hardcoded `libgcc` with library name extracted from
path, so we are compatible with both libraries.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

semihalf-duda-patryk commented:
I changed the original idea a bit. Now, the path where linker should search (`-L`) and library name (`-l`) is extracted from `--print-libgcc-file-name`, so we are compatible with both runtime libraries.

duda-patryk committed:
clang: Provide --target option when determining path to runtime library

Clang/LLVM is natively a cross-compiler, so one set of applications can
compile code to all supported targets. The default target can be changed
using '--target' option.

CMake supports this type of compilers. To change compiling target, one
should set CMAKE_C_COMPILER_TARGET accorgindly.

The '--target' option has impact on the path to clang-rt library
returned by compiler when run with '--print-libgcc-file-name' option.
Without specifying target, Clang will return path to runtime library of
the host target (e.g. x86_64-pc-linux-gnu).

Signed-off-by: Patryk Duda <pdk@semihalf.com>

semihalf-duda-patryk commented:
When `CMAKE_C_COMPILER_TARGET` is set, CMake internally adds the `-target` option when invoking the compiler, but if we invoke the compiler manually, we need to provide the `-target` option, otherwise compiler will return wrong path to runtime library.

dcpleung commented:
Hm... so even if you ask for `libgcc`, the compiler returns rtlib?

semihalf-duda-patryk commented:
> Hm... so even if you ask for `libgcc`, the compiler returns rtlib?

I think, my description is misleading. Clang provides correct path to chosen library, some examples from my machine:

```
clang --rtlib=libgcc --print-libgcc-file-name
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/libgcc.a
```
```
clang --rtlib=compiler-rt --print-libgcc-file-name
/usr/lib64/clang/17/lib/linux/libclang_rt.builtins-x86_64.a
```
```
clang --target=armv7m-none-eabi --rtlib=compiler-rt --print-libgcc-file-name
/usr/lib64/clang/17/lib/baremetal/libclang_rt.builtins-armv7m.a
```
```
clang --target=armv7m-none-eabi --rtlib=libgcc --print-libgcc-file-name
libgcc.a # (probably because I don't have libgcc for arm)
```

If we don't provide `--target` option, we will always get path to library default target (x86_64 usually).
